### ISSUE TITLE
Question about prepare_aligned_crop

### ISSUE THREAD

**Issue by shuangshuangguo** [xwjabc/hed#4]
Hi, Thanks for your sharing! Could you tell me why to use 'prepare_aligned_crop' function in building network. I do not understand.

**Comment by xwjabc** [comment]
Hi! The prepare_aligned_crop function is used to mimic the crop function in Caffe. Assume a point P(x,y) in the input image X. If we add a Conv2D (kernel 3x3, padding 1) over the input image, the corresponding spatial position of original P on Conv2D(X) will still be P'(x,y). If we add a bilinear upsampling layer (stride=2) over the input image, the corresponding spatial position of original P on Upsample(X) will be P''(2x, 2y). Thus, the prepare_aligned_crop function will try to calculate (a rough) point-to-point linear mapping between the input image and side output in order to guide the crop process (which is used in Caffe crop layer). Actually, you can also directly apply center cropping for side output, but here I just hope to fully reproduce the behavior of the official HED code.

**Comment by shuangshuangguo** [comment]
Sorry, I do not understand. Did you mean the crop layer is used to calculate linear mapping?
From the following code, I think the crop layer computes crop start point. Am I right?

![image](https://user-images.githubusercontent.com/31126075/51434747-f6a81100-1ca2-11e9-8c8f-1d0ac632ffda.png)

**Comment by xwjabc** [comment]
Yes. We will obtain the mapping `x' = ax + b` at first. Actually, after the upsampling, `a=1`. Thus, the mapping will be `x' = x + b` and the cropping start position will be `b`.
You can refer to the code below: (`score_dsn1_map, upsample2_map, upsample3_map, upsample4_map, upsample5_map` contain the corresponding mapping coefficient (a, b).)
```
        crop1_margin = int(score_dsn1_map[1])
        crop2_margin = int(upsample2_map[1])
        crop3_margin = int(upsample3_map[1])
        crop4_margin = int(upsample4_map[1])
        crop5_margin = int(upsample5_map[1])
```

**Comment by shuangshuangguo** [comment]
I think I have understood what happened.
The unsampled image has different size with the corresponding input image, so we should find the coordinate of original point (0, 0) in unsampled image to implement cropping.

**Comment by xwjabc** [comment]
Yes. Your understanding is correct.

**Comment by shuangshuangguo** [comment]
Thank you very much! 
By the way, the code seems not adapt to multi-gpu training. If someone want to use multi-gpu, he should modify torch.nn.functional.conv_transpose2d to torch.nn.ConvTranspose2d. Am I right?

**Comment by xwjabc** [comment]
Hi. Currently, the code does not support mini-batch > 1 or multi-GPU training because it requires input images to have the same size (since we directly use the original images and their sizes are varied). If you still need multi-GPU training, you may need to rewrite the DataParallel part to support various sizes of input images.

**Comment by shuangshuangguo** [comment]
I have rewrote the code to support various sizes of input images. But I still can not use multi-gpu because self.weight_deconv2(3, 4, 5) is placed on GPU0 only. Now I try to modify the networks.py.

**Comment by xwjabc** [comment]
I see. If that is the case, you are right that you can directly use `torch.nn.ConvTranspose2d` instead of current way.

**Comment by shuangshuangguo** [comment]
Thank you!

**Comment by herleeyandi** [comment]
@xwjabc Hello I still don't understand the `prepare_aligned_crop` do. Suppose I want to use the `resnet` as feature extractor that using pretrained model, and I add several `bottleneck layer` for predicting the edges. Should I do `map_compose` for every convolution layer in `bottleneck layer` and also the residual block in resnet layer?

**Comment by xwjabc** [comment]
I think you can simply use center crop instead of aligned crop. Previously I used aligned crop in order to fully reproduce the performance of HED (by mimicking the Caffe crop layers). Actually, center crop will have similar results.

**Comment by herleeyandi** [comment]
@xwjabc Sorry I am a beginner in here, could you be more specific about the center crop?,  do you mean that even without mimicking caffe crop layer, the result will be the same which means the prepare align crop function like conv_map, pool_map, ETC is not necessary?

I tried to use another feature extractor, and use the similar HED architecture as side output, which means for every stage in feature extractor followed by the hed conv, for example, 1st stage with `conv1_2`, `conv2_2`, that similar in original HED, and only HED part is trained and the feature extractor will be freeze,  but I got this result.
![batch-27999-1st-image](https://user-images.githubusercontent.com/2309538/54181352-3a680b80-44d9-11e9-97db-2c326d802948.png)

Could you suggest what should I do?, thank you so much.

**Comment by xwjabc** [comment]
1) Yes, in networks.py, I used:
```
        # Aligned cropping.
        crop1 = score_dsn1[:, :, self.crop1_margin:self.crop1_margin+image_h,
                                 self.crop1_margin:self.crop1_margin+image_w]
        crop2 = upsample2[:, :, self.crop2_margin:self.crop2_margin+image_h,
                                self.crop2_margin:self.crop2_margin+image_w]
        crop3 = upsample3[:, :, self.crop3_margin:self.crop3_margin+image_h,
                                self.crop3_margin:self.crop3_margin+image_w]
        crop4 = upsample4[:, :, self.crop4_margin:self.crop4_margin+image_h,
                                self.crop4_margin:self.crop4_margin+image_w]
        crop5 = upsample5[:, :, self.crop5_margin:self.crop5_margin+image_h,
                                self.crop5_margin:self.crop5_margin+image_w]
```
which is based on aligned cropping. Center crop means using the center part (size: `image_h, image_w`) of `score_dsn1`, `upsample2`, `upsample3`, `upsample4`, `upsample5` as final crops.

2) First, you may need to make the feature extractor part trainable. If it does not work, you may try to lower down the learning rate.

**Comment by herleeyandi** [comment]
@xwjabc Yes I finally use the center crop that fixing the width to crop. Anyway, I find that this crop is to make sure the feature map is in the same size, therefore the crops can be concatenated and aligned, am I right?
For my previous problem the reason are the bad weight initialization and training paramaters. I just find the better ones. and I get this result.
![batch-6499-1st-image](https://user-images.githubusercontent.com/2309538/54228186-944ded00-453c-11e9-82ef-e94d74670ff2.png)
However I still confuse with the last weighted binary cross entrophy. why it is not in categorical?, I just curious for the last convolution why it is `self.score_final = nn.Conv2d(5, 1, 1)` not `self.score_final = nn.Conv2d(5, 2, 1)` ?, since we will classify whether it is edge or non-edge right?, it should be 2 and we can use pytorch native BCE loss. Why you just output it to 1 channel?

**Comment by xwjabc** [comment]
The output 1 channel indicates the probability of the edge (0: non-edge / 1: edge).
You can refer to `def weighted_cross_entropy_loss(preds, edges)` in `hed.py` about the use of BCE loss.

**Comment by GaoyangLiu** [comment]
> @xwjabc Yes I finally use the center crop that fixing the width to crop. Anyway, I find that this crop is to make sure the feature map is in the same size, therefore the crops can be concatenated and aligned, am I right?
> For my previous problem the reason are the bad weight initialization and training paramaters. I just find the better ones. and I get this result.
> ![batch-6499-1st-image](https://user-images.githubusercontent.com/2309538/54228186-944ded00-453c-11e9-82ef-e94d74670ff2.png)
> However I still confuse with the last weighted binary cross entrophy. why it is not in categorical?, I just curious for the last convolution why it is `self.score_final = nn.Conv2d(5, 1, 1)` not `self.score_final = nn.Conv2d(5, 2, 1)` ?, since we will classify whether it is edge or non-edge right?, it should be 2 and we can use pytorch native BCE loss. Why you just output it to 1 channel?

I met a similar problem, where the predictions are all gray, white, or black. Could u tell me how u solve it? What weight initialization and training parameters u took? Thank u very much.
